incubator/hnc/pkg/reconcilers/hierarchical_namespace.go

```diff
@@ -59,51 +59,97 @@ func (r *HierarchicalNamespaceReconciler) Reconcile(req ctrl.Request) (ctrl.Resu
 	nm := req.Name
 	pnm := req.Namespace
 
-	// Get instance from apiserver. If the instance doesn't exist, we don't want to reconcile
-	// it since it may trigger the HC reconciler to recreate the namespace that was just deleted.
-	// TODO expand on this to check the owner's hc.spec.allowCascadingDelete. If it's set to
-	//  true, we still want to reconcile the hns instance. See issue:
-	//  https://github.com/kubernetes-sigs/multi-tenancy/issues/501
+	// Get instance from apiserver. If the instance doesn't exist, we don't want
+	// to reconcile it to create the HNS instance or the namespace.
 	inst, err := r.getInstance(ctx, pnm, nm)
 	if err != nil {
 		if errors.IsNotFound(err) {
-			// If the instance doesn't exist, return nil to prevent a retry.
 			return ctrl.Result{}, nil
 		}
 		return ctrl.Result{}, err
 	}
 
-	// Report "Forbidden" state and early exist if the namespace is not allowed to self-serve
-	// namespaces but has bypassed the webhook and successfully created the hns instance.
+	// Report "Forbidden" state and early exist if the namespace is not allowed to
+	// self-serve namespaces but has bypassed the webhook and successfully created
+	// the hns instance. Forbidden HNSes won't have finalizers.
 	// TODO refactor/split the EX map for 1) reconciler exclusion and 2) self-serve not allowed
 	//  purposes. See issue: https://github.com/kubernetes-sigs/multi-tenancy/issues/495
 	if EX[pnm] {
 		inst.Status.State = api.Forbidden
 		return ctrl.Result{}, r.writeInstance(ctx, log, inst)
 	}
 
-	// Get the self-serve namespace's hierarchyConfig and namespace instances.
-	nsInst, err := r.getNamespace(ctx, nm)
+	// Get the owned child namespace instance. If it doesn't exist, initialize one.
+	cnsInst, err := r.getNamespace(ctx, nm)
 	if err != nil {
 		return ctrl.Result{}, err
 	}
 
-	switch {
-	case nsInst.Name == "":
-		log.Info("The owned subnamespace does not exist", "subnamespace", nm)
-		inst.Status.State = api.Missing
+	// If the instance is being deleted, remove the finalizers or cascading delete.
+	if !inst.DeletionTimestamp.IsZero() {
+		if len(inst.ObjectMeta.Finalizers) == 0 {
+			// We've finished process this, nothing to do.
+			return ctrl.Result{}, nil
+		}
+		if deleteNamespace := r.deleting(log, inst, cnsInst); deleteNamespace {
+			r.deleteNamespace(ctx, log, cnsInst)
+		}
+		return ctrl.Result{}, r.writeInstance(ctx, log, inst)
+	}
+
+	r.updateState(log, inst, cnsInst)
+	if inst.Status.State == api.Missing {
 		if err := r.writeNamespace(ctx, log, nm, pnm); err != nil {
 			return ctrl.Result{}, err
 		}
-	case nsInst.Annotations[api.AnnotationOwner] != pnm:
-		log.Info("The owner annotation of the subnamespace doesn't match the owner", "annotation", nsInst.Annotations[api.AnnotationOwner])
+	}
+
+	// Add finalizers on all non-forbidden HNSes.
```

Comment:
Maybe say _why_ we're adding the finalizer (e.g. to ensure it's not deleted until after the child namespace is deleted)